libbeat/common/transport/httpcommon/proxy_uri.go

```diff
@@ -0,0 +1,109 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package httpcommon
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// ProxyURI is a URL used for proxy serialized as a string.
+type ProxyURI url.URL
+
+func NewProxyURIFromString(s string) (*ProxyURI, error) {
+	u, err := url.Parse(s)
+	if err != nil {
+		return &ProxyURI{}, err
+	}
+
+	return NewProxyURIFromURL(*u), nil
+}
+
+func NewProxyURIFromURL(u url.URL) *ProxyURI {
+	p := ProxyURI(u)
+	return &p
+}
+
+// MarshalYAML serializes URI as a string.
+func (p ProxyURI) MarshalYAML() (interface{}, error) {
```

Comment:
`ProxyURI` is stored by pointer in the config struct. Will the yaml marshaller find that method? Reflection can be a pain at times, reflection rules are not as forgiving about pointers vs. non-pointers as the compiler. Consider to use pointers more consistently.